teps/0124-distributed-tracing-for-tasks-and-pipelines.md

```diff
@@ -0,0 +1,72 @@
+---
+status: proposed
+title: Distributed tracing for Tasks and Pipelines
+creation-date: '2022-09-30'
+last-updated: '2022-09-30'
+authors:
+- '@kmjayadeep'
+---
+
+# TEP-0124: Distributed tracing for Tasks and Pipelines
+
+<!-- toc -->
+- [Summary](#summary)
+  - [Goals](#goals)
+  - [Non-Goals](#non-goals)
+  - [Use Cases](#use-cases)
+- [Proposal](#proposal)
+  - [PipelineRun controller](#pipelinerun-controller)
+  - [TaskRun controller](#taskrun-controller)
+  - [Test Plan](#test-plan)
+<!-- /toc -->
+
+## Summary
+
+With distributed tracing, we can track the time taken by each action in the pipeline like reconciling logic, fetching resources, pulling images etc.
+This allows the developers to improve the reconciliation logic and also allow end users to monitor and optimize the pipelines.
+
+### Goals
+
+* Implementation of opentelemetry tracing with Jaeger
+* Instrumentation of pipelinerun and taskrun reconciliation logic
+* Able to visualize pipeline and task reconciliation steps in jaeger
+
+### Non-Goals
+
+* Instrumentation of sidecars and initcontainers
+* Support for more tracing backends
+* Instrumentation of individual steps in each task
+* Adding events to each span to indicate what is happening inside each
+  method. This is an improvement task, and it can be done later once the
+  basic setup for tracing is in place. The scope of this proposal is
+  only to get the plumbing done by covering only the method boundaries.
+
+### Use Cases
+
+Pipeline and Task User:
+* I would like to understand the duration of each task in my pipeline so that I can optimize the slow taks to improve the pipeline execution speed
+
+Tekton Developer:
+* I would like to understand the duration of each reconciliation step, so that I can optimize the code to improve reconciliation performance
+* When the pipelines are failing due to a bug, I would like to understand which reconciliation logic caused the issue so that I can easily fix the problem
+
+## Proposal
+
+Initialize a tracer provider with jaeger as the backend for each reconciler. The jaeger collector URL can be passed as an argument to the controller.
```

Comment:
Passing the URL as an argument to the controller should be enough for development environments.
Eventually, it can be an enhancement out of scope for this initial TEP, we might want to implement reading configuration from a config map, so that configuration may be altered at runtime, similar to https://pkg.go.dev/knative.dev/pkg/tracing. `knative/pkg` relies on OpenCensus for now, and we would like to to start using OpenTelemetry right from the beginning.